tools/load-generator/main.go

```diff
@@ -0,0 +1,239 @@
+
+	for {
+		start := time.Now()
+
+		for _, query := range q.queries {
+			q.query(query.Expr)
+		}
+
+		wait := q.interval - time.Since(start)
+		if wait > 0 {
+			time.Sleep(wait)
+		}
+	}
+}
+
+func (q *Querier) query(expr string) {
+	queryCount.WithLabelValues(q.target, q.name, expr, q.qtype).Inc()
+	start := time.Now()
+
+	req, err := http.NewRequest("GET", q.url, nil)
+	if err != nil {
+		log.Printf("Error creating request: %v", err)
+		queryFailCount.WithLabelValues(q.target, q.name, expr, q.qtype).Inc()
+		return
+	}
+
+	qParams := req.URL.Query()
+	qParams.Set("query", expr)
+	if q.qtype == "range" {
+		qParams.Set("start", fmt.Sprintf("%d", int64(time.Now().Add(-q.start).Unix())))
+		qParams.Set("end", fmt.Sprintf("%d", int64(time.Now().Add(-q.end).Unix())))
+		qParams.Set("step", q.step)
+	}
+	req.URL.RawQuery = qParams.Encode()
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		log.Printf("Error querying Prometheus: %v", err)
+		queryFailCount.WithLabelValues(q.target, q.name, expr, q.qtype).Inc()
+		return
+	}
+	defer resp.Body.Close()
+
+	duration := time.Since(start)
+	queryDuration.WithLabelValues(q.target, q.name, expr, q.qtype).Observe(duration.Seconds())
+
+	if resp.StatusCode == http.StatusNotFound {
+		log.Printf("WARNING: GroupID#%d: Querier returned 404 for Prometheus instance %s.", q.groupID, q.url)
+		q.numberOfErrors++
+		if q.numberOfErrors >= max404Errors {
+			log.Fatalf("ERROR: GroupID#%d: Querier returned 404 for Prometheus instance %s %d times.", q.groupID, q.url, max404Errors)
+		}
+	} else if resp.StatusCode != http.StatusOK {
+		log.Printf("WARNING: GroupID#%d: Querier returned %d for Prometheus instance %s.", q.groupID, resp.StatusCode, q.url)
+	} else {
+		body, _ := io.ReadAll(resp.Body)
+		log.Printf("GroupID#%d: query %s %s, status=%d, size=%d, duration=%.3f", q.groupID, q.target, expr, resp.StatusCode, len(body), duration.Seconds())
+	}
+}
+
+func durationSeconds(s string) time.Duration {
+	num := s[:len(s)-1]
+	value, err := time.ParseDuration(num + string(s[len(s)-1]))
+	if err != nil {
+		log.Fatalf("Invalid duration: %s", s)
+	}
+	return value
+}
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Println("unexpected arguments")
+		fmt.Println("usage: <load_generator> <namespace> <pr_number>")
+		os.Exit(2)
+	}
+	prNumber := os.Args[2]
+
+	configFile, err := os.ReadFile("/etc/loadgen/config.yaml")
+	if err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+
+	var config struct {
+		Querier struct {
+			Groups []QueryGroup `yaml:"groups"`
+		} `yaml:"querier"`
+	}
+	if err := yaml.Unmarshal(configFile, &config); err != nil {
+		log.Fatalf("Failed to parse config: %v", err)
+	}
+
+	fmt.Println("Loaded configuration")
+
+	var wg sync.WaitGroup
+
+	for i, group := range config.Querier.Groups {
+		wg.Add(1)
+		go NewQuerier(i, "pr", prNumber, group).run(&wg)
+	}
+
+	for i, group := range config.Querier.Groups {
+		wg.Add(1)
+		go NewQuerier(i, "release", prNumber, group).run(&wg)
+	}
```

Comment:
This can be done in one loop.

```suggestion
	for i, group := range config.Querier.Groups {
		wg.Add(1)
		go NewQuerier(i, "pr", prNumber, group).run(&wg)
		wg.Add(1)
		go NewQuerier(i, "release", prNumber, group).run(&wg)
	}
```